src/backends/gpu_kernels.hpp

```diff
@@ -0,0 +1,434 @@
+
+        // backward sweep
+        for(auto i=last_max; i>first; i-=BlockWidth) {
+            if (i<=last) {
+                auto factor = u[i] / d[i];
+                d[p[i]]   -= factor * u[i];
+                rhs[p[i]] -= factor * rhs[i];
+            }
+        }
+
+        __syncthreads();
+        rhs[first] /= d[first];
+
+        // forward sweep
+        for(auto i=first+BlockWidth; i<=last; i+=BlockWidth) {
+            rhs[i] -= u[i] * rhs[p[i]];
+            rhs[i] /= d[i];
+        }
+    }
+}
+
+// Data is to be interleaved into blocks of width BlockWidth.
+// The kernel assigns LoadWidth threads to each lane in the block.
+// Hence each thread block is responsible for loading a single block
+// of interleaved matrices.
+
+template <typename T, typename I, int BlockWidth, int LoadWidth, int Threads>
+__global__
+void interleave(
+    const T* in, T* out, const I* sizes, const I* starts, int padded_size, int num_mtx)
+{
+    static_assert(BlockWidth*LoadWidth==Threads, "");
+
+    __shared__ T buffer[Threads];
+
+    const auto tid = threadIdx.x + blockIdx.x*blockDim.x;
+    const auto lid = threadIdx.x;
+
+    const auto mtx_id   = tid/LoadWidth;
+    const auto mtx_lane = tid - mtx_id*LoadWidth;
+
+    const auto blk_id   = tid/(BlockWidth*LoadWidth);
+    const auto blk_row  = lid/BlockWidth;
+    const auto blk_lane = lid - blk_row*BlockWidth;
+
+    const auto blk_pos  = LoadWidth*blk_lane + blk_row;
+
+    const bool do_load  = mtx_id<num_mtx;
+
+    // only threads that participate in loading access starts and sizes arrays
+    auto load_pos  = do_load? starts[mtx_id] + mtx_lane     : 0;
+    const auto end = do_load? starts[mtx_id] + sizes[mtx_id]: 0;
+    auto store_pos = blk_id*BlockWidth*padded_size + (blk_row*BlockWidth + blk_lane);
+
+    for (auto i=0; i<padded_size; i+=LoadWidth) {
+        auto loaded = impl::npos<T>();
+        if (do_load && load_pos<end) {
+            loaded = in[load_pos];
+        }
+        buffer[lid] = loaded;
+        __syncthreads();
+        if (i+blk_row<padded_size) {
+            out[store_pos] = buffer[blk_pos];
+        }
+        load_pos  += LoadWidth;
+        store_pos += LoadWidth*BlockWidth;
+    }
+}
+
+// host side wrapper for the interleave kernel
+template <typename T, typename I, int BlockWidth, int LoadWidth>
+void interleave(const T* in, T* out, const I* sizes, const I* starts, int padded_size, int num_mtx)
+{
+    constexpr int Threads = BlockWidth*LoadWidth;
+    const int blocks = impl::block_count(num_mtx, BlockWidth);
+
+    interleave<T, I, BlockWidth, LoadWidth, Threads>
+        <<<blocks, Threads>>> (in, out, sizes, starts, padded_size, num_mtx);
+}
+
+// A helper that performs the interleave operation on host memory.
+template <typename T, typename I>
+std::vector<T> interleave_host(
+        const std::vector<T>& in,
+        const std::vector<I>& sizes,
+        const std::vector<I>& starts,
+        int block_width, int num_mtx, int padded_length)
+{
```

Comment:
It is only used internally inside the interlave matrix constructor, and as a helper in one of the unit tests. So, I would rather not have a uniform interface + wrapper for such limited (and non-user facing) uses.